feat(ai-gateway): AI Gateway 2.0 entities (#5057)
Kong/developer.konghq.com: app/_ai_gateway_entities/agent.md

      send subsequent requests. Rewriting the `url` field, and any `additionalInterfaces[].url`
      fields, to the {{site.ai_gateway}} address means clients route follow-up traffic through the
      gateway instead of bypassing it. The rewrite honors `X-Forwarded-*` headers when the gateway
      sits behind a load balancer.

  - q: How does streaming work?
    a: |
      Server-sent events (`Content-Type: text/event-stream`) pass through chunk-by-chunk without
      buffering. The runtime counts SSE events, captures time-to-first-byte, and extracts task state
      from the final event for analytics. Latency is preserved.

  - q: How do I limit which consumers can reach an Agent?
    a: |
      Set the `acls` field on the Agent with allow or deny lists. Each entry is a string that
      references a Consumer, Consumer Group, or Authenticated Group by name.

  - q: Can the same plugin run on an Agent that I'd attach to a route or service?
    a: |
      Plugin configuration that applies to the Agent goes through the [Policy entity](/ai-gateway/entities/policy/).
      Attach Policies to the Agent through its `policies` field.
---

## What is an Agent?

> An Agent is a first-class {{site.ai_gateway}} entity that represents an upstream agent endpoint exposed through {{site.ai_gateway}}. An Agent has a type, either `a2a` for [Agent-to-Agent protocol](https://a2aproject.github.io/A2A/) traffic or `http` for generic HTTP agent routing, plus configuration that points {{site.ai_gateway}} at the upstream and shapes how requests flow.

Reviewer: ```suggestion
An Agent is a first-class {{site.ai_gateway}} entity that represents an upstream agent endpoint exposed through {{site.ai_gateway}}. An Agent has a type, either `a2a` for [Agent-to-Agent protocol](https://a2aproject.github.io/A2A/) traffic or `http` for generic HTTP agent routing, and a configuration that points {{site.ai_gateway}} at the upstream and shapes how requests flow.
```

> For `a2a` Agents, the runtime adds protocol-aware behavior on top of plain proxying: it detects A2A requests across both JSON-RPC and REST bindings, rewrites agent-card URLs so clients discover the gateway as the canonical endpoint, and emits structured A2A telemetry to {{site.konnect_short_name}} analytics and OpenTelemetry. For `http` Agents, requests are proxied without A2A-specific processing.

Reviewer: Start with the basic case then follow with the enhanced version.

```suggestion
For `http` type Agents, requests are proxied without A2A-specific processing. For `a2a` type Agents, {{site.ai_gateway}} adds protocol-aware behaviour on top of plain proxying: it detects A2A requests across both JSON-RPC and REST bindings, rewrites agent-card URLs so clients discover the gateway as the canonical endpoint, and emits structured A2A telemetry to {{site.konnect_short_name}} analytics and OpenTelemetry. 
```

> Agents can be created and managed through {{site.konnect_short_name}}, the on-prem Admin API, decK, or the {{site.konnect_short_name}} UI:

Reviewer: Mentions Konnect twice?

Reviewer 2: Fixed

{% table %}
columns:
  - title: Deployment
    key: deployment
  - title: Control Plane
    key: cp
  - title: Endpoint
    key: endpoint
rows:
  - deployment: "{{site.konnect_short_name}}"
    cp: "{{site.konnect_short_name}} {{site.ai_gateway}} API"
    endpoint: /v1/ai-gateways/{aiGatewayId}/agents
  - deployment: On-prem
    cp: Admin API
    endpoint: /ai/agents
{% endtable %}

## How A2A traffic flows

> When an Agent has type `a2a`, the runtime processes traffic in four phases:

Reviewer: There's a lot of places where you say 'the runtime does X', I think it would be better to say 'AI Gateway does X' in general or to distinguish the 'dataplane/controlplane/AI proxy plugin does X' when important. Alternatively avoid having a specific subject of the sentence.

```suggestion
When an Agent has type `a2a`, proxied traffic is processed in four phases:
```

> 1. **Access**. Detects whether the request is an A2A operation (JSON-RPC or REST binding). When statistics logging is enabled, starts an OpenTelemetry span and records the request body for payload logging if that's enabled too.

Reviewer: ```suggestion
1. **Access**. Detects whether the request is an A2A operation (JSON-RPC or REST binding). When statistics logging is enabled, this starts an OpenTelemetry span and records the request body for payload logging if that's also enabled.
```

Non-A2A traffic, and traffic to `http` Agents, is proxied without these steps.

<!-- vale off -->
{% mermaid %}
sequenceDiagram
    autonumber
    participant Client as A2A Client
    participant Gateway as {{site.ai_gateway}}<br>(Agent)
    participant Agent as Upstream A2A Agent

    Client->>Gateway: A2A request (JSON-RPC or REST)
    Note over Gateway: Detect A2A binding and method<br>Start OTel span (if logging enabled)

    Gateway->>Agent: Proxied request<br>(Accept-Encoding removed if logging enabled)

    alt Streaming response (SSE)
        Agent-->>Gateway: text/event-stream chunks
        Note over Gateway: Pass through each chunk<br>Count SSE events, track TTFB
        Gateway-->>Client: SSE chunks (unchanged)
        Note over Gateway: On final chunk:<br>Extract task state, set analytics
    else Non-streaming response
        Agent->>Gateway: JSON response
        Note over Gateway: Buffer response<br>Extract task metadata
        Gateway->>Client: Response (unchanged)
    end

    Note over Gateway: Finish OTel span<br>Emit ai.a2a metrics to log plugins
{% endmermaid %}
<!-- vale on -->

> ## Core A2A protocol elements

Reviewer: Not an immediate concern, but I think we could split this doc in two and have a seperate A2A concepts doc.

Reviewer 2: yeah, let's do it in the second iteration

    purpose: Enables clients to discover agents and understand how to interact with them.
  - element: Task
    description: A stateful unit of work initiated by an agent, with a unique ID and defined lifecycle.
    purpose: Tracks long-running operations and supports multi-turn interactions.
  - element: Message
    description: A single turn of communication between a client and an agent, containing content and a role (`user` or `agent`).
    purpose: Conveys instructions, context, questions, answers, or status updates that are not formal artifacts.
  - element: Part
    description: The fundamental content container (for example, `TextPart`, `FilePart`, `DataPart`) used within messages and artifacts.
    purpose: Provides flexibility for agents to exchange different content types within messages and artifacts.
  - element: Artifact
    description: A tangible output generated by an agent during a task (for example, a document, image, or structured data).
    purpose: Carries the concrete output of a task in a structured, retrievable form.
{% endtable %}

## Protocol detection

> A2A traffic is auto-detected per request. There's no per-route opt-in, and non-A2A traffic passes through without overhead.

Reviewer: ```suggestion
A2A traffic is auto-detected per request and non-A2A traffic passes through without overhead.
```

> **REST binding.** Detection anchors to the end of the request path, so any prefix added by the route is ignored. For example, both `/v1/message:send` and `/api/agents/v1/message:send` match `SendMessage`:

Reviewer: Why is this `<strong>` ?

Reviewer 2: fixed

    key: canonical
rows:
  - method: "`POST`"
    path: "`/v1/message:send`"
    operation: SendMessage
    canonical: "`message/send`"
  - method: "`POST`"
    path: "`/v1/message:stream`"
    operation: SendStreamingMessage
    canonical: "`message/stream`"
  - method: "`GET`"
    path: "`/.well-known/agent-card.json`"
    operation: GetAgentCard
    canonical: "`agent/getCard`"
  - method: "`GET`"
    path: "`/v1/extendedAgentCard`"
    operation: GetExtendedAgentCard
    canonical: "`agent/getExtendedAgentCard`"
  - method: "`GET`"
    path: "`/v1/tasks/{id}`"
    operation: GetTask
    canonical: "`tasks/get`"
  - method: "`GET`"
    path: "`/v1/tasks`"
    operation: ListTasks
    canonical: "`tasks/list`"
  - method: "`POST`"
    path: "`/v1/tasks/{id}:cancel`"
    operation: CancelTask
    canonical: "`tasks/cancel`"
  - method: "`POST`"
    path: "`/v1/tasks/{id}:subscribe`"
    operation: SubscribeToTask
    canonical: "`tasks/resubscribe`"
  - method: "`POST`"
    path: "`/v1/tasks`"
    operation: ListTasks
    canonical: "`tasks/list`"
{% endtable %}
<!-- vale on -->

The canonical method name is what appears in OpenTelemetry span attributes and log output.

> **JSON-RPC binding.** Detected by the `"jsonrpc"` field in the request body, combined with a recognized A2A method name or an `A2A-Version` request header. Recognized methods include `message/send`, `message/stream`, `tasks/get`, `tasks/list`, `tasks/cancel`, `tasks/resubscribe`, the `tasks/pushNotificationConfig/*` family, and `agent/getExtendedAgentCard`.

Reviewer: same as above, the use of `<strong>` here seems out of place, this fits better as a H3

Reviewer 2: fixed

A request carrying an `A2A-Version` header is treated as JSON-RPC even if the method isn't in the recognized list. When an unknown method is accepted this way, the `method` field in log output is recorded as `"unknown"` to bound metric cardinality. The OpenTelemetry span's `kong.a2a.operation` attribute still receives the actual method name.

> ## Agent-card URL rewriting

Reviewer: Consider nesting 'Protocol detection' and 'Agent-card URL rewriting' under 'Core A2A protocol elements'

Reviewer 2: fixed

When an upstream agent returns an agent card, the runtime rewrites the `url` field, and any `additionalInterfaces[].url` fields, to the {{site.ai_gateway}} address. A2A clients then discover the gateway as the canonical endpoint instead of contacting the upstream directly. The rewrite uses `X-Forwarded-*` headers to construct the correct scheme, host, and port when the gateway is deployed behind a load balancer or reverse proxy.

## Logging and observability

> Statistics logging records structured A2A telemetry per request: the A2A method, binding type, task state, task ID, context ID, latency, time to first byte (for streaming), SSE event count, and response size. The runtime emits this data into the `ai.a2a` namespace consumed by {{site.konnect_short_name}} analytics and any attached log plugins, and creates a `kong.a2a` child span when {{site.base_gateway}} tracing is configured.

Reviewer: ```suggestion
When Statistics logging is enabled the {{site.ai_gateway}} records the following structured A2A telemetry per request: 

- A2A method
- Binding type
- Task state
- Task ID
- Context ID
- Latency
- Time to first byte (for streaming)
- SSE event count
- Response size. 

The runtime emits this data into the `ai.a2a` namespace consumed by {{site.konnect_short_name}} analytics and any attached logging plugins, and creates a `kong.a2a` child span when {{site.base_gateway}} tracing is configured.
```

Reviewer: As a rule I would not have inline lists longer that 3 elements, these become very hard to read.

{:.info}
> When statistics logging is enabled, the runtime removes the `Accept-Encoding` request header
> before forwarding to the upstream. This prevents compressed responses that the runtime can't
> parse for metadata extraction.

> Payload logging additionally captures request and response bodies. Payloads are truncated at the configured payload size limit. Enable with care. Payload logging may expose sensitive data.

Reviewer: Don't say 'Enable with care' inline. Either this is important enough for a `{:.warning}` callout or it can be removed.

Reviewer 2: yeah, it was a sloppy note of mine from a rough draft

You can view A2A analytics in {{site.konnect_short_name}} Explorer and Dashboards through the [Agentic usage analytics](/observability/explorer/?tab=agentic-usage#metrics) view.

### Log output fields

{% include /plugins/ai-a2a-proxy/log-output-fields.md %}

### OpenTelemetry span attributes

> When statistics logging is enabled and {{site.base_gateway}} tracing is configured, the runtime creates a `kong.a2a` child span with the following attributes:

Reviewer: Where is tracing configured?
